### DISCUSSION TITLE
Invite Permission

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#13365] [Archive]
Where can i find the settings for the permission to be able to invite a user?

**Comment by azrikahar** [comment]
This is not configurable at the moment, but do you have any suggestions on how it would look like from your perspective?

**Reply by eikaramba** [reply]
from my point of view the most important thing is to be able to validate the role attribute. currently it seems this can be a big security risk if one sets the role on the frontend. as far as i can see it cannot be validated. if a filter hook or permission can be set than this can at least be validated on the server before the email is send out.

**Reply by NilsBaumgartner1994** [reply]
> from my point of view the most important thing is to be able to validate the role attribute. currently it seems this can be a big security risk if one sets the role on the frontend. as far as i can see it cannot be validated. if a filter hook or permission can be set than this can at least be validated on the server before the email is send out.



> from my point of view the most important thing is to be able to validate the role attribute. currently it seems this can be a big security risk if one sets the role on the frontend. as far as i can see it cannot be validated. if a filter hook or permission can be set than this can at least be validated on the server before the email is send out.

Please open a new topic for this

**Reply by eikaramba** [reply]
but isn't that exactly this topic here? i mean we are talking about the same setting roles->permissions -> invite permission(not there yet) - there you would do this too.

**Reply by NilsBaumgartner1994** [reply]
For me it seems like an extra topic

**Reply by rijkvanzanten** [reply]
> from my point of view the most important thing is to be able to validate the role attribute. currently it seems this can be a big security risk if one sets the role on the frontend. as far as i can see it cannot be validated. if a filter hook or permission can be set than this can at least be validated on the server before the email is send out.

The permissions for inviting a person are the same as for creating a directus_users item. You can configure the role validation on there 👍🏻

**Comment by NilsTessier** [comment]
for me it seems good enough  to do it by setting permissions on directus_users for the Public role like that :

![drectus_userfieldpermissions](https://user-images.githubusercontent.com/50356564/181539659-7aab8c67-3ee8-4397-a953-88c8c252d065.png)

for the field permissions i found that only the email , role and status fields are really needed. 
Since we don't want users to self elevate or self activate we need to set field validation like that: 

![directus_user_validation](https://user-images.githubusercontent.com/50356564/181540291-da2bb0ff-a89a-42ed-a3a1-c197f32c00f4.png)

if somebody catch a security issue that I wouldn't have thought of please let me know

**Reply by rijkvanzanten** [reply]
That should do the trick 👍🏻 Alternatively, you can remove write access to the role field completely, in favor of using a preset to set the default role

**Reply by manuelwehrli** [reply]
> That should do the trick 👍🏻 Alternatively, you can remove write access to the role field completely, in favor of using a preset to set the default role

Is this really possible with field presets? I tried it and get only errors that the role field is required in the post request or that it has to match the uuid of an existing role. My idea was that the field preset value will be set as role. Removing create permission (if you mean that with write access), I get even permission denied errors.

**Reply by rijkvanzanten** [reply]
If you add `{ "role": "<id of your 'basic user' role>" }` as the preset under _Field Presets_, you should be able to remove the field permission for the role field 👍🏻

**Reply by manuelwehrli** [reply]
> 

I tried many things now, I gave permissions to more and sometimes less fields, in the most minimal case I just gave create permission for email and password, though password is actually only required for the second step which is accepting the invitation over the link. Using the invite endpoint I pass an email, then it says a role is required, so I pass a role with a valid uuid too, but then it just says I do not have permission. What else can I try? Does the basic role also need any create permission for users? Is your hint only for the /users endpoint or should it work the same way with the /users/invite endpoint (as you stated somewhere they base on the same permissions) so newcomers can publicly self-invite which is kind of the best "public regristration" that directus has at the moment? Funny fact, when passing the token of an admin user with the exact same request, it creates a user, but without the given email address nor a role assigned. Is this a bug?

**Reply by rijkvanzanten** [reply]
What I'm saying is for the `/users` endpoint. The `/users/invite` endpoint reuses the same permissions as users, though it does require the role to be submitted https://github.com/directus/directus/blob/36c824e364f1acfec3c74097724bd92dbd4cf5d0/api/src/controllers/users.ts#L273-L277

**Reply by eikaramba** [reply]
@manuelwehrli you can also try to enable all system collections permissions. for an app i implemented the invitation workflow. However i reimplemented it myself basically, but i cannot remember why. i think it was because of some issues with the email i had and wanted to do some custom stuff. Maybe the invite_url also needs to be whitelisted

**Reply by manuelwehrli** [reply]
Thanks for the hints. The /users endpoint works now as it should. Is it somehow possible to trigger the invite routine (which nicely includes generating one time jwt and sending the email) with a flow? This because using the invite endpoint publicly seems impossible without providing a role in the payload and public create permission on role and status, which allows then to (ab)use /users and create random invited users directly. (Everything would be so much easier if we could set permissions separately for the invite endpoint.) Can I at least use some function to generate that jwt of the invitation link inside a flow? I also wondered if flows are able to define a create hook (filter) that prevents the actual creation of an item (responding with an error) if a promise in the flow is rejected, or if this first api action always happens and I can just do things in addition to it? So would it mean that to implement a kind of custom register/invitation, I need a an extra collection that temporarily stores the email address (public create permission) and then triggering the whole flow, which might check the email against common spam/blacklists, and in the end creates a new user with invited status? Or even better, the flow would create only an email with a jwt link that allows the recipient to create a new user with that email address (rather than changing an invited one to active), this way avoiding that an earlier attempt to register with the same address gives an error of duplicate email in the users collection? There would neither be any temporary data to clean up and no extra collection is needed. I hope you can guide me in the best direction to solve public registration with a combination of all the most recent directus features, avoiding whenever possible writing custom endpoints.
